R/load_survey.r

```diff
@@ -73,44 +72,37 @@ load_survey <- function(files, ...) {
     can_merge <- vapply(survey_files, function(x) {
       length(intersect(colnames(contact_data[[x]]), colnames(main_surveys[[type]]))) > 0
     }, TRUE)
-    merge_files <- names(can_merge[which(can_merge)])
+    merge_files <- names(can_merge[can_merge])
     while (length(merge_files) > 0) {
       merged_files <- NULL
       for (file in merge_files)
       {
         common_id <- intersect(colnames(contact_data[[file]]), colnames(main_surveys[[type]]))
 
-        ## try a merge to see whether it can be done uniquely
-        test_merge <- merge(
-          main_surveys[[type]], contact_data[[file]], by = common_id,
-          all.x = TRUE, allow.cartesian = TRUE
-        )
-
-        if (nrow(test_merge) == nrow(main_surveys[[type]])) {
-          ## check if all IDs can be merged in
-          unique_main_survey_ids <-
-            unique(main_surveys[[type]][, common_id, with = FALSE])
-          unique_additional_survey_ids <-
-            unique(contact_data[[file]][, common_id, with = FALSE])
-
-          id_overlap <- merge(
-            unique_main_survey_ids, unique_additional_survey_ids,
-            by = common_id
+        # is the id unique and can the merge be done uniquely?
+        if (anyDuplicated(main_surveys[[type]][, common_id, with = FALSE]) == 0) {
+          test_merge <- merge(
+            main_surveys[[type]], contact_data[[file]], by = common_id,
+            all.x = TRUE
           )
 
-          if (nrow(id_overlap) < nrow(unique_main_survey_ids)) {
+          if (nrow(test_merge) > nrow(contact_data[[file]])) {
```

Comment:
This is not the right test for: unique ids in the main survey that are not present in the file that is to be merged (the point of this test). I think you'll need to revert to the previous version where `id_overlap` is first created and then checked (not a very expensive operation).

Reply:
Your approach fails to flag any problem if `contact_data[[file]]` happens to have the same number of rows as `main_surveys[[type]]` but the IDs don't overlap. It's probably such an unlikely case that perhaps it's not worth the extra `data.table::merge` call, but it is currently happening in the UK CoMix survey because of the stray `X` columns (see https://github.com/epiforecasts/socialmixr/issues/74) (i.e. won't happen any more once that is fixed).